README.md
# What is Locker?

Locker is a simple web application for claiming and releasing locks.
It contains a Concourse resource to help make locking various tasks
and jobs in Concourse pipelines easier to manage.

This project is similar to the [pool-resource](https://github.com/concourse/pool-resource), with a few key differences:

1) It uses a webserver with file-backed persistence for the locks. This isn't
   terribly scalable, but since it's handling locks for Concourse pipelines,
   it isn't anticipated to have an enormous traffic load.
2) Locks and Keys do not need to be pre-defined in order to be used
3) You may lock a Lock multiple times with the same Key. If this is done,
   you must unlock once for each lock in place on the key.
4) Each lock request may specify who it is, along with the key. If not specified,
   the value of the key will be used as the identifier. This allows users to lock
   a resource that can be shared by node-a, node-b, and node-c, but not node-e.

# How do I use it?

1) Deploy `locker` along side your Concourse database node via the [locker-boshrelease](https://github.com/cloudfoundry-community/locker-boshrelease)
2) Use the [locker-resource](https://github.com/cloudfoundry-community/locker-resource) in your Concourse pipelines.

# locker API

## Supported Requests

* `GET /locks`

  Returns a JSON formatted list of locks + who owns them currently

* `PUT /lock/<lock-name>`

  Content: `{"key":"key-to-lock-with", "lock_by": "identifier-of-lock-requestor"}`

  Issues a lock on `lock-name` keyed with `key` attribute in the JSON payload of the request.
  If the lock was already taken, it will immediately return a 423 error, and the client should back-off +
  re-try at a sane interval until the lock is obtained. The `lock_by` field is optional, and used to
  identify the item requesting the lock. If not specified, it will default to the value of `key`.

  `locked_by` allows you to lock a lock multiple times with a single key, by multiple items. Different
  keys should be specified by items that need exclusivity on the lock. If three things can run simultaneously,
  but the fourth item must be exclusive, use two keys, and one key should have three `locked_by` requestors.

  Returns 200 on success, 423 on locking failure

  Example to lock `prod-deployments` with `prod-cloudfoundry`:

  ```
  curl -X PUT -d '{"key":"prod-cloudfoundry"}' http://locker-ip:port/lock/prod-deployments
  # or to explicitly set the lock requestor
  curl -X PUT -d '{"key":"prod-cloudfoundry", "lock_by": "my-deployment"}' http://locker-ip:port/lock/prod-deployments
  ```

* `DELETE /lock/<lock-name>`

  Content: `{"key":"key-to-unlock-with"}`

  Issues an unlock request on `lock-name` using the `key` attribute in the JSON payload as the unlock key.
  If the lock on `lock-name` is not currently keyed by `key-to-unlock-with`, the
  unlock is disallowed. If the lock is currently not held by anyone, always returns 200.

  Returns 200 on success, 423 on failure.

  Example to unlock `prod-deployments` previously locked with `prod-cloudfoundry`:

  ```
  curl -X DELETE -d '{"key":"prod-cloudfoundry"}' http://locker-ip:port/lock/prod-deployments
  # or to unlock my-deployment's lock using `prod-cloudfoundry`
  curl -X DELETE -d '{"key":"prod-cloudfoundry","locked_by":"my-deployment"}' http://locker-ip:port/lock/prod-deployments
  ```

## Authentication

`locker` is protected with HTTP basic authentication. Credentials are passed in 

## Error handling

Errors will be reported as json objects, inside a top-level `error` key:

```
{"error":"This is what went wrong"}
```

## Running manually

If you want to run `locker` manually for testing/development:

```
go build
LOCKER_CONFIG=/tmp/locker-data.yml ./locker
```

`locker` can be configured using the following environment variables:

* `LOCKER_CONFIG` **required** - specifies the file that locks will be stored in
* `PORT` - Defaults to 3000, controls the port `locker` listens on
* `AUTH_USER` - If specified, requires `AUTH_PASS` and configures the username for
  HTTP basic auth
* `AUTH_PASS` - If specified, requires `AUTH_USER` and configures the password for
  HTTP basic auth


ci/release_notes.md
# Improvements

A single unlock request for a key/requestor now results in all locks
for that key/requestor pair to be removed. A single key/requestor can still
unlock multiple times without error, but after the first one, its locks
have been released, and subsequent unlocks are no-ops. If another requestor
still retains a lock, the key is not removed.

This should make it easier to recover from failures, as you can now request
a lock, fail, request a lock again, succeed, release your lock, and not need
to intervene manually to clean up the original lock.


locker_test.go
package main_test

import (
	"io/ioutil"

	. "github.com/cloudfoundry-community/locker"
	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"
)

var _ = Describe("Locker", func() {
	var locker Locker
	BeforeEach(func() {
		tmpFile, err := ioutil.TempFile("", "locker")
		Expect(err).ShouldNot(HaveOccurred())

		locker = Locker{LockConfig: tmpFile.Name()}
	})
	Context("when locking", func() {
		It("uses the key as requestor if requestor is not provided", func() {
			err := locker.Lock("myLock", "myKey", "")
			Expect(err).ShouldNot(HaveOccurred())
			lock, err := locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.Key).Should(Equal("myKey"))
			Expect(lock.LockedBy).Should(Equal(map[string]int{"myKey": 1}))
		})
		It("increments requestor when the key is correct", func() {
			err := locker.Lock("myLock", "myKey", "")
			Expect(err).ShouldNot(HaveOccurred())
			err = locker.Lock("myLock", "myKey", "")
			Expect(err).ShouldNot(HaveOccurred())
			lock, err := locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.Key).Should(Equal("myKey"))
			Expect(lock.LockedBy).Should(Equal(map[string]int{"myKey": 2}))
		})
		It("fails to lock, if the key doesn't match what's in the lock", func() {
			err := locker.Lock("myLock", "myKey", "")
			Expect(err).ShouldNot(HaveOccurred())
			err = locker.Lock("myLock", "newKey", "")
			Expect(err).Should(HaveOccurred())
			lock, err := locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.Key).Should(Equal("myKey"))
			Expect(lock.LockedBy).Should(Equal(map[string]int{"myKey": 1}))
		})
		It("is able to lock multiple requestors multiple times, if they all have the correct key", func() {
			err := locker.Lock("myLock", "myKey", "george")
			Expect(err).ShouldNot(HaveOccurred())
			err = locker.Lock("myLock", "myKey", "jane")
			Expect(err).ShouldNot(HaveOccurred())
			err = locker.Lock("myLock", "myKey", "george")
			Expect(err).ShouldNot(HaveOccurred())
			err = locker.Lock("myLock", "myKey", "jane")
			Expect(err).ShouldNot(HaveOccurred())

			lock, err := locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.Key).Should(Equal("myKey"))
			Expect(lock.LockedBy).Should(Equal(map[string]int{
				"jane":   2,
				"george": 2,
			}))
		})
	})
	Context("when unlocking", func() {
		It("uses the key as requestor if requestor is not provided", func() {
			err := locker.Lock("myLock", "myKey", "")
			Expect(err).ShouldNot(HaveOccurred())
			err = locker.Unlock("myLock", "myKey", "")
			Expect(err).ShouldNot(HaveOccurred())
			lock, err := locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.Key).Should(Equal(""))
			Expect(lock.LockedBy).Should(Equal(map[string]int{"myKey": 0}))
		})
		It("fully unlocks a key/requestor on a single unlock call", func() {
			err := locker.Lock("myLock", "myKey", "george")
			Expect(err).ShouldNot(HaveOccurred())
			err = locker.Lock("myLock", "myKey", "george")
			Expect(err).ShouldNot(HaveOccurred())
			lock, err := locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.LockedBy["george"]).Should(Equal(2))

			err = locker.Unlock("myLock", "myKey", "george")
			Expect(err).ShouldNot(HaveOccurred())
			lock, err = locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.LockedBy["george"]).Should(Equal(0))
		})
		It("does not error when decrementing a requestor that is not currently holding a lock", func() {
			err := locker.Lock("myLock", "myKey", "george")
			Expect(err).ShouldNot(HaveOccurred())
			err = locker.Unlock("myLock", "myKey", "jane")
			Expect(err).ShouldNot(HaveOccurred())

			lock, err := locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.LockedBy["george"]).Should(Equal(1))
			Expect(lock.LockedBy["jane"]).Should(Equal(0))
		})
		It("does not error when no one is holding a lock", func() {
			lock, err := locker.GetLock("myMissingLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock).Should(Equal(Lock{LockedBy: map[string]int{}}))
			err = locker.Unlock("myMissingLock", "anyKey", "whoCares?")
			Expect(err).ShouldNot(HaveOccurred())
			lock, err = locker.GetLock("myMissingLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock).Should(Equal(Lock{Key: "", LockedBy: map[string]int{"whoCares?": 0}}))
		})
		It("removes the key from the lock, if no remaining requestors hold locks", func() {
			err := locker.Lock("myLock", "myKey", "george")
			Expect(err).ShouldNot(HaveOccurred())
			lock, err := locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.Key).Should(Equal("myKey"))
			err = locker.Unlock("myLock", "myKey", "george")
			Expect(err).ShouldNot(HaveOccurred())
			lock, err = locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.Key).Should(Equal(""))
		})
		It("fails to unlock, if the key doesn't match what's in the lock", func() {
			err := locker.Lock("myLock", "myKey", "george")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(locker.Unlock("myLock", "wrongKey", "george")).ShouldNot(Succeed())
		})
		It("unlocking one of multiple requestors does not remove the key frmo the lock", func() {
			Expect(locker.Lock("myLock", "myKey", "george")).Should(Succeed())
			Expect(locker.Lock("myLock", "myKey", "george")).Should(Succeed())
			Expect(locker.Lock("myLock", "myKey", "jane")).Should(Succeed())
			Expect(locker.Lock("myLock", "myKey", "jane")).Should(Succeed())
			Expect(locker.Unlock("myLock", "myKey", "george")).Should(Succeed())
			lock, err := locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.LockedBy).Should(Equal(map[string]int{"jane": 2, "george": 0}))
		})
	})
	Context("when loading", func() {
		It("throws an error when the locker data could not be loaded", func() {
			locker.LockConfig = "DoesNotExist"
			_, err := locker.GetLocks()
			Expect(err).Should(HaveOccurred())
		})
		It("returns the on-disk locker data", func() {
			err := ioutil.WriteFile(locker.LockConfig, []byte(`
{
	"myLock": {
		"key": "myKey",
		"locked_by": {
			"george": 2,
			"jane": 15
		}
	}
}
`), 0644)
			Expect(err).ShouldNot(HaveOccurred())
			lock, err := locker.GetLock("myLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.Key).Should(Equal("myKey"))
			Expect(lock.LockedBy).Should(Equal(map[string]int{"george": 2, "jane": 15}))
		})
	})
	Context("when saving", func() {
		It("throws an error when the locker data could not be saved", func() {
			locker.LockConfig = "Should/Fail/To/Save"
			err := locker.SaveLocks(LockerState{"myLock": Lock{Key: "myKey", LockedBy: map[string]int{"george": 2}}})
			Expect(err).Should(HaveOccurred())
		})
		It("saves locker data to disk", func() {
			err := locker.SaveLocks(LockerState{"myLock": Lock{Key: "myKey", LockedBy: map[string]int{"george": 2}}})
			Expect(err).ShouldNot(HaveOccurred())
			data, err := ioutil.ReadFile(locker.LockConfig)
			Expect(err).ShouldNot(HaveOccurred())
			Expect(string(data)).Should(MatchJSON(`{"myLock":{"key":"myKey","locked_by":{"george":2}}}`))
		})
	})
	Context("when retrieving a lock", func() {
		It("returns the requested lock", func() {
			Expect(locker.Lock("myLock", "myKey", "george")).Should(Succeed())
			Expect(locker.Lock("otherLock", "otherKey", "jane")).Should(Succeed())
			lock, err := locker.GetLock("otherLock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock).Should(Equal(Lock{Key: "otherKey", LockedBy: map[string]int{"jane": 1}}))
		})
		It("auto-vivifies the LockedBy map", func() {
			lock, err := locker.GetLock("notYetALock")
			Expect(err).ShouldNot(HaveOccurred())
			Expect(lock.LockedBy).ShouldNot(BeNil())
			Expect(lock.LockedBy).Should(Equal(map[string]int{}))
		})
	})
})


locker_suite_test.go
package main_test

import (
	"testing"

	. "github.com/onsi/ginkgo"
	. "github.com/onsi/gomega"
)

func TestLocker(t *testing.T) {
	RegisterFailHandler(Fail)
	RunSpecs(t, "Locker Test Suite")
}


locker.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
)

type LockerState map[string]Lock

type Locker struct {
	LockConfig string
}

type Lock struct {
	LockedBy map[string]int `json:"locked_by"`
	Key      string         `json:"key"`
}

func (l Locker) GetLocks() (LockerState, error) {
	locks := LockerState{}

	data, err := ioutil.ReadFile(l.LockConfig)
	if err != nil {
		return nil, err
	}
	if len(data) == 0 {
		return locks, nil
	}
	err = json.Unmarshal(data, &locks)
	if err != nil {
		return nil, err
	}
	return locks, nil
}
func (l Locker) SaveLocks(locks LockerState) error {
	data, err := json.MarshalIndent(locks, "", "    ")
	if err != nil {
		return err
	}
	err = ioutil.WriteFile(l.LockConfig, data, 0644)
	if err != nil {
		return err
	}
	return nil
}
func (l Locker) GetLock(pool string) (Lock, error) {
	locks, err := l.GetLocks()
	if err != nil {
		return Lock{}, err
	}
	lock := locks[pool]
	if lock.LockedBy == nil {
		lock.LockedBy = map[string]int{}
	}
	return lock, nil
}
func (l Locker) SetLock(pool string, lock Lock) error {
	locks, err := l.GetLocks()
	if err != nil {
		return err
	}
	locks[pool] = lock
	return l.SaveLocks(locks)
}
func (l Locker) Lock(pool, key, requestor string) error {
	lock, err := l.GetLock(pool)
	if err != nil {
		return err
	}
	if requestor == "" {
		requestor = key
	}
	if lock.Key != key && lock.Key != "" {
		return fmt.Errorf("Attempt to steal lock for '%s' with '%s' by '%s' thwarted. Currently held by someone else", pool, key, requestor)
	}
	lock.LockedBy[requestor] += 1
	lock.Key = key
	return l.SetLock(pool, lock)
}
func (l Locker) Unlock(pool, key, requestor string) error {
	lock, err := l.GetLock(pool)
	if err != nil {
		return err
	}
	if requestor == "" {
		requestor = key
	}
	if lock.Key != key && lock.Key != "" {
		return fmt.Errorf("Attempt to unlock '%s' with '%s' by '%s' thwarted. Currently held by someone else", pool, key, requestor)
	}

	lock.LockedBy[requestor] = 0
	totalLocks := 0
	for _, locks := range lock.LockedBy {
		totalLocks += locks
	}
	// reset the key if we are totally unlocked
	if totalLocks == 0 {
		lock.Key = ""
	}
	return l.SetLock(pool, lock)
}


server.go
package main

import (
	"fmt"
	"os"
)

type LockOperation string

const (
	LockOp   LockOperation = "Lock"
	UnlockOp               = "Unlock"
	ListOp                 = "ListLocks"
)

type LockStatus int

const (
	Locked LockStatus = iota
	Unlocked
	Error
	None
)

type LockRequest struct {
	Command  LockOperation
	Pool     string
	Lock     LockInput
	Response chan LockResponse
}
type LockResponse struct {
	Status  LockStatus
	Message interface{}
	Error   error
}

type LockInput struct {
	Key       string `json:"key"`
	Requestor string `json:"requestor"`
}

func lockServer(lockRequests chan LockRequest, lockConfig string) {
	locker := Locker{LockConfig: lockConfig}
	_, err := locker.GetLocks()
	if err != nil {
		panic(fmt.Sprintf("Unable to load lock config from '%s': %s", lockConfig, err))
	}
	for req := range lockRequests {
		if req.Command == ListOp {
			locks, err := locker.GetLocks()

			res := LockResponse{
				Status:  None,
				Message: locks,
				Error:   err,
			}
			req.Response <- res
		} else if req.Command == LockOp {
			res := LockResponse{}

			current, err := locker.GetLock(req.Pool)
			if err != nil {
				res.Status = Error
				res.Error = err
				req.Response <- res
				continue
			}

			err = locker.Lock(req.Pool, req.Lock.Key, req.Lock.Requestor)
			if err != nil {
				res.Status = Error
				res.Error = err
				req.Response <- res
				continue
			}
			current, err = locker.GetLock(req.Pool)
			if err != nil {
				res.Status = Error
				res.Error = err
				req.Response <- res
				continue
			}
			if current.Key != req.Lock.Key {
				res.Status = Error
				res.Error = fmt.Errorf("Locking failed. Should be locked by '%s', but found '%s'", req.Lock.Key, current.Key)
				req.Response <- res
				continue
			}

			res.Status = Locked
			res.Message = map[string]string{
				"response": fmt.Sprintf("Lock for '%s' acquired by '%s' using key '%s'", req.Pool, req.Lock.Requestor, req.Lock.Key),
			}
			req.Response <- res
		} else if req.Command == UnlockOp {
			res := LockResponse{}

			err := locker.Unlock(req.Pool, req.Lock.Key, req.Lock.Requestor)
			if err != nil {
				res.Status = Error
				res.Error = err
				req.Response <- res
				continue
			}

			res.Status = Unlocked
			res.Message = map[string]string{"response": fmt.Sprintf("'%s' released a lock on on '%s'", req.Lock.Requestor, req.Pool)}
			req.Response <- res
		} else {
			fmt.Fprintf(os.Stderr, "Invalid lock request '%s'", req.Command)
			req.Response <- LockResponse{Error: fmt.Errorf("Invalid lock request '%s'", req.Command)}
		}
	}
}


main.go
package main

import (
	"encoding/base64"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/gin-gonic/gin/render"
	"github.com/voxelbrain/goptions"
)

var Version string

func main() {
	var options struct {
		Help    bool `goptions:"-h, --help"`
		Version bool `goptions:"-v, --version"`
	}

	goptions.ParseAndFail(&options)
	if options.Help {
		goptions.PrintHelp()
		os.Exit(0)
	}
	if options.Version {
		if Version != "" {
			fmt.Fprintf(os.Stderr, "locker v%s\n", Version)
		} else {
			fmt.Fprintf(os.Stderr, "locker (development build)\n")
		}
		os.Exit(0)
	}

	lockChan := make(chan LockRequest)
	lockConfig := os.Getenv("LOCKER_CONFIG")
	if lockConfig == "" {
		fmt.Fprintf(os.Stderr, "No LOCKER_CONFIG file specified. Bailing out.\n")
		os.Exit(1)
	}
	_, err := os.Stat(lockConfig)
	if os.IsNotExist(err) {
		err := ioutil.WriteFile(lockConfig, []byte("{}"), 0644)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Unable to create LOCKER_CONFIG file (%s): %s\n", lockConfig, err)
			fmt.Fprintf(os.Stderr, "Bailing out.\n")
			os.Exit(1)
		}
	}

	authUser := os.Getenv("AUTH_USER")
	authPass := os.Getenv("AUTH_PASS")
	if authUser != "" && authPass == "" {
		fmt.Fprintf(os.Stderr, "AUTH_USER specified, but no AUTH_PASS was provided. Bailing out.\n")
		os.Exit(1)
	}
	if authUser == "" && authPass != "" {
		fmt.Fprintf(os.Stderr, "AUTH_PASS specified, but no AUTH_USER was provided. Bailing out.\n")
		os.Exit(1)
	}

	certFile := os.Getenv("SSL_CERT")
	keyFile := os.Getenv("SSL_KEY")
	if certFile == "" && keyFile != "" {
		fmt.Fprintf(os.Stderr, "SSL_KEY specified, but no SSL_CERT was provided. Bailing out.\n")
		os.Exit(1)
	}
	if certFile != "" && keyFile == "" {
		fmt.Fprintf(os.Stderr, "SSL_CERT specified, but no SSL_KEY was provided. Bailing out.\n")
		os.Exit(1)
	}

	go lockServer(lockChan, lockConfig)

	router := gin.Default()
	var routes *gin.RouterGroup
	if authUser != "" && authPass != "" {
		routes = router.Group("/", AuthHandler(authUser, authPass))
	} else {
		routes = router.Group("/", func(c *gin.Context) {})
	}
	routes.GET("/locks", func(c *gin.Context) {
		rc := make(chan LockResponse)

		lr := LockRequest{
			Command:  ListOp,
			Response: rc,
		}

		lockChan <- lr

		response := <-rc
		if response.Error != nil {
			fmt.Fprintf(os.Stderr, "Error listing locks: %s\n", response.Error)
			c.Render(500, render.JSON{Data: map[string]string{"error": response.Error.Error()}})
			return
		}

		c.Render(200, render.JSON{Data: response.Message})
	})
	routes.PUT("/lock/:pool", func(c *gin.Context) {
		rc := make(chan LockResponse)

		pool := c.Param("pool")
		input := LockInput{}
		c.BindJSON(&input)
		if input.Key == "" {
			c.Render(400, render.JSON{Data: map[string]string{"error": "No key specified in request body"}})
			return
		}
		if input.Requestor == "" {
			input.Requestor = input.Key
		}

		lr := LockRequest{
			Command:  LockOp,
			Pool:     pool,
			Lock:     input,
			Response: rc,
		}
		lockChan <- lr
		res := <-rc
		if res.Error != nil {
			fmt.Fprintf(os.Stderr, "Error locking %s with %s: %s\n", pool, input.Key, res.Error)
			c.Render(423, render.JSON{Data: map[string]string{"error": res.Error.Error()}})
			return
		}

		if res.Status == Locked {
			c.Render(200, render.JSON{Data: res.Message})
		} else {
			fmt.Fprintf(os.Stderr, "Unexpected response to a lock request: %#v\n", res)
			c.Render(500, render.JSON{Data: map[string]string{"error": "Unknown response from lock request"}})
		}
	})
	routes.DELETE("/lock/:pool", func(c *gin.Context) {
		rc := make(chan LockResponse)
		pool := c.Param("pool")
		input := LockInput{}
		c.BindJSON(&input)
		if input.Key == "" {
			c.Render(400, render.JSON{Data: map[string]string{"error": "No lock specified in request body"}})
			return
		}
		if input.Requestor == "" {
			input.Requestor = input.Key
		}

		lr := LockRequest{
			Command:  UnlockOp,
			Pool:     pool,
			Lock:     input,
			Response: rc,
		}

		lockChan <- lr
		res := <-rc
		if res.Error != nil {
			fmt.Fprintf(os.Stderr, "Error unlocking %s: %s\n", pool, res.Error)
			c.Render(423, render.JSON{Data: map[string]string{"error": res.Error.Error()}})
			return
		}

		if res.Status != Unlocked {
			fmt.Fprintf(os.Stderr, "Unexpected response to unlock request: %#v\n", res)
			c.Render(500, render.JSON{Data: map[string]string{"error": "Unknown response from unlock request"}})
			return
		}

		c.Render(200, render.JSON{Data: res.Message})
	})

	port := os.Getenv("PORT")
	if port == "" {
		port = "3000"
	}
	if certFile == "" && keyFile == "" {
		err := http.ListenAndServe(":"+port, router)
		fmt.Fprintf(os.Stderr, "Error running webserver: %s\n", err)
		os.Exit(1)
	} else if certFile != "" && keyFile != "" {
		err := http.ListenAndServeTLS(":"+port, certFile, keyFile, router)
		fmt.Fprintf(os.Stderr, "Error running webserver: %s\n", err)
		os.Exit(1)
	}
	fmt.Fprintf(os.Stderr, "Unable to choose TLS or non-TLS web server. Bailing out\n")
	os.Exit(0)
}

func AuthHandler(user, password string) gin.HandlerFunc {
	realm := "Basic realm=\"Authorization Required\""
	return func(c *gin.Context) {
		hdr := c.Request.Header.Get("Authorization")
		fields := strings.SplitN(hdr, " ", 2)
		if strings.ToLower(fields[0]) != "basic" {
			fmt.Fprintf(os.Stderr, "Got illegal authentication request type: %s\n", fields[0])
			c.Header("WWW-Authenticate", realm)
			c.Render(401, render.JSON{Data: map[string]string{"error": "Please Authenticate"}})
			c.Abort()
			return
		}

		auth, err := base64.StdEncoding.DecodeString(fields[1])
		if err != nil {
			fmt.Fprintf(os.Stderr, "Failed to decode authorization header '%s': %s", fields[1], err)
			c.Header("WWW-Authenticate", realm)
			c.Render(401, render.JSON{Data: map[string]string{"error": "Please Authenticate"}})
			c.Abort()
			return
		}
		creds := strings.SplitN(string(auth), ":", 2)
		if user != creds[0] || password != creds[1] {
			c.Header("WWW-Authenticate", realm)
			c.Render(401, render.JSON{Data: map[string]string{"error": "Please Authenticate"}})
			c.Abort()
			return
		}
	}
}
